### PR TITLE
feat(cheatcodes): add ability to ignore (multiple) specific and partial reverts in fuzz and invariant tests

### LINKED ISSUE #4271: feat(forge): ability to exclude certain custom errors and revert reason strings from failing tests
### Component

Forge

### Describe the feature you would like

The use case is fork + fuzz tests - certain tokens like USDC have blacklists that prevent the transfer of assets between certain accounts (see lines 837 and 838 [here](https://etherscan.io/address/0xa2327a938febf5fec13bacfb16ae10ecbc4cbdcf#code)). When Foundry fuzzes a blacklisted account, the Forge test will fail because of the following error:

> Blacklistable: account is blacklisted

However, this is a false negative. The user's test should not have failed.

It would be great if there were two config options for filtering revert reason strings and custom errors.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by emo-eth**
## Motivation

As raised in #4271, sometimes the fuzzer dictionary will uncover real but irrelevant errors in contract code, especially when testing against stateful forks. Runs that encounter anticipated errors should be thrown out and count towards the global `assume` counter.

Closes #4271.

## Solution

This PR adds several overloaded and related function definitions to `vm.assumeNoRevert`, introduced in #8780, as follows:

```solidity
    /// Discard this run's fuzz inputs and generate new ones if next call reverted with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoPartialRevert`.
    #[cheatcode(group = Testing, safety = Safe)]
    function assumeNoRevert(bytes4 revertData) external pure;

    /// Discard this run's fuzz inputs and generate new ones if next call reverted with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoPartialRevert`.
    #[cheatcode(group = Testing, safety = Safe)]
    function assumeNoRevert(bytes calldata revertData) external pure;

    /// Discard this run's fuzz inputs and generate new ones if next call reverted with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoRevert`.
    #[cheatcode(group = Testing, safety = Safe)]
    function assumeNoRevert(bytes4 revertData, address reverter) external pure;

    /// Discard this run's fuzz inputs and generate new ones if next call reverted with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoPartialRevert`.
    #[cheatcode(group = Testing, safety = Safe)]
    function assumeNoRevert(bytes calldata revertData, address reverter) external pure;

    /// Discard this run's fuzz inputs and generate new ones if next call reverted with data starting with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoRevert`.
    #[cheatcode(group = Testing, safety = Safe)]
    function assumeNoPartialRevert(bytes4 revertData) external pure;

    /// Discard this run's fuzz inputs and generate new ones if next call reverted with data starting with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoRevert`.
    #[cheatcode(group = Testing, safety = Safe)]
    function assumeNoPartialRevert(bytes4 revertData, address reverter) external pure;
```

The `reverter` param ensures a revert comes from a specific called contract, per #8770, and the `PartialRevert` methods allow matching on selectors with arbitrary extra data, per #8763.

### Behavior
- it should be possible to to reject any specific error without data (`vm.assumeNoRevert(bytes4)`)
- it should be possible to reject any specific error with arbitrary data (`vm.assumeNoRevert(bytes)`)
- it should be possible to reject any specific of error regardless of extra data (`vm.assumeNoPartialRevert(bytes4)`)
- it should be possible to reject multiple errors that a single call may produce (multiple subsequent calls to any of the `vm.assumeNoRevert` definitions that accept params)
- it should be possible to restrict these rejections to a specific reverter (the optional overloaded `address reverter` param)
- anticipated reverts should reset after next non-cheatcode external call
- combining a catch-all `vm.assumeNoRevert()` with a `vm.assumeNoRevert(args)` call should result in an error

### Questions

- Should these functions be `pure` and `Safe`? I see that `expectRevert` is not `pure`, and is marked as `Unsafe`, but `assumeNoRevert()` is marked `pure` and `Safe`.
- Is the behavior of calling `vm.assumeNoRevert` multiple times to build up anticipated reverts per-call acceptable, or should it be forced to take an array of structs with revert parameters, since I think aggregating parameters from separate calls is new behavior (correct me if I'm wrong).
- I've also left a few `TODO`s in the code with questions, which I will try to surface via comments here.
- `expectRevert()` without args seems to not play strangely with cheatcode errors – eg, `vm.expectRevert()` will swallow a cheatcode revert, including those thrown by `vm.assumeNoRevert` and subsequent calls to `vm.expectRevert()`. This seems like a separate bug that I can log if we agree there.

### CODE REVIEW

File: crates/cheatcodes/src/test/revert.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
+use crate::{Error, Result};
```
Comment (emo-eth) [inline crates/cheatcodes/src/test/revert.rs:1]: Is a pub(crate) module the best place for this shared logic?

File: crates/cheatcodes/src/inspector.rs [hunk 2]
```diff
@@ -753,6 +753,9 @@ where {
                     outcome.result.result,
                     outcome.result.output.clone(),
                     &self.config.available_artifacts,
+                    // todo: is this specific to `forge create` calls? only tx.sender (equivalent
+                    // to tx.origin?) is set; it would be wrong to say tx sender is the reverter
+                    None,
```
Comment (emo-eth) [inline crates/cheatcodes/src/inspector.rs:758]: Curious if `None` is the correct "reverter" here

File: crates/cheatcodes/src/test/assume.rs [hunk 3]
```diff
@@ -20,10 +63,153 @@ impl Cheatcode for assumeCall {
...
+    known_contracts: &Option<ContractsByArtifact>,
+    reverter: Option<&Address>,
+) -> Result<()> {
+    // iterate over acceptable reasons and try to match against any, otherwise, return an Error with
+    // the revert data
+    assume_no_revert.reasons.as_ref().map_or_else(
+        || {
+            // todo: fix this hack to get cheatcode name in error message
+            let retdata_str = retdata.to_string();
+            if retdata_str.contains(ASSUME_REJECT_MAGIC) ||
```
Comment (emo-eth) [inline crates/cheatcodes/src/test/assume.rs:196]: I'm packing for a trip tomorrow so will take a look tomorrow afternoon, but for now `testMultipleAssumes_ThrowOnGenericNoRevert_fails` fails because `vm.assumeNoRevert` is swallowing the cheatcode revert from a more-specific call to `vm.assumeNoRevert`. `apply_dispatch` seems to modify the revert data (after checking it's a string), so this check is probably failing that comparison.

File: crates/cheatcodes/src/inspector.rs (hunk 2 above)
Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:758]: The `create_end` interpreter callback is called whenever a contract is created, so it is also called in tests when for example you do a `new Counter()`; Agree sender should not be marked as reverter, so None is the right choice.
I don't think we need extra reverter parameter in `handle_expect_revert` fn, the address that reverts tx is already recorded (or None) in `ExpectedRevert.reverted_by` and we already pass the expected revert

File: crates/cheatcodes/src/inspector.rs [hunk 4]
```diff
@@ -1252,15 +1255,45 @@ impl Inspector<&mut dyn DatabaseExt> for Cheatcodes {
...
+                // specific reason was supplied
                 if outcome.result.is_revert() {
-                    outcome.result.output = Error::from(MAGIC_ASSUME).abi_encode().into();
+                    let assume_no_revert = std::mem::take(&mut self.assume_no_revert).unwrap();
+                    return match handle_assume_no_revert(
+                        &assume_no_revert,
+                        outcome.result.result,
+                        &outcome.result.output,
+                        &self.config.available_artifacts,
+                        assume_no_revert.reverted_by.as_ref(),
```
Comment (grandizzy) [inline crates/cheatcodes/src/inspector.rs:1275]: same as above, no need for extra `reverter` param as we can retrieve itr in ` handle_assume_no_revert` as `assume_no_revert.reverted_by.as_ref()`

File: crates/cheatcodes/src/test/revert.rs (hunk 1 above)
Comment (grandizzy) [reply crates/cheatcodes/src/test/revert.rs:1]: If I am not wrong the intention was to have names per cheatcodes (assume contains logic for `vm.assume*` cheatcodes, expect for `vm.expect*` cheatcodes) so revert name could be confusing in this context (though not sure what a good name could be, maybe handlers?).
If we add such probably we should move `handle_assume_no_revert` here too.

File: crates/forge/tests/cli/test_cmd.rs [hunk 5]
```diff
@@ -2390,3 +2391,260 @@ contract Dummy {
 
     assert!(dump_path.exists());
 });
+
+forgetest_init!(assume_no_revert_tests, |prj, cmd| {
```
Comment (grandizzy) [inline crates/forge/tests/cli/test_cmd.rs:2395]: for new cheatcodes tests are added in `testdata/default/cheats` and picked up by `test_cheats_*` tests, please move them there
https://github.com/foundry-rs/foundry/blob/9fe891ab5babbdc2891c67d14d6c75ea1ca4b19c/crates/forge/tests/it/cheats.rs#L14-L16

see sample https://github.com/foundry-rs/foundry/blob/master/testdata/default/cheats/ExpectRevert.t.sol

**Review by grandizzy — COMMENTED** [review]
looks good! left couple of comments, going to move back in draft until fix added

**Comment by grandizzy** [comment]
@mds1 would be great to have your thoughts re new proposed cheatcodes. Thanks!

File: crates/forge/tests/cli/test_cmd.rs (hunk 5 above)
Comment (grandizzy) [reply crates/forge/tests/cli/test_cmd.rs:2395]: 👍 yep, those can stay here

File: crates/cheatcodes/src/test/revert.rs (hunk 1 above)
Comment (emo-eth) [reply crates/cheatcodes/src/test/revert.rs]: @grandizzy Changed the file to `revert_handlers.rs` and moved the handler fns in here – lmk what you think!

**Comment by emo-eth** [comment]
@grandizzy should be in a good state now – apologies, had a hectic week. let me know if anything else looks off.

**Comment by jenpaff** [comment]
> @grandizzy should be in a good state now – apologies, had a hectic week. let me know if anything else looks off.

moving PR to ready for review

File: crates/cheatcodes/spec/src/vm.rs [hunk 6]
```diff
@@ -825,6 +825,30 @@ interface Vm {
...
+
+    /// Discard this run's fuzz inputs and generate new ones if next call reverted with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoRevert`.
+    #[cheatcode(group = Testing, safety = Safe)]
+    function assumeNoRevert(bytes4 revertData, address reverter) external pure;
+
+    /// Discard this run's fuzz inputs and generate new ones if next call reverted with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoPartialRevert`.
+    #[cheatcode(group = Testing, safety = Safe)]
+    function assumeNoRevert(bytes calldata revertData, address reverter) external pure;
+
+    /// Discard this run's fuzz inputs and generate new ones if next call reverted with data starting with the given revert data. Call more than once to add multiple reasons. Can be combined with `assumeNoRevert`.
```
Comment (grandizzy) [inline crates/cheatcodes/spec/src/vm.rs:844]: I think we can get rid of `function assumeNoPartialRevert` cheatcodes, the `assumeParatialRevert` counterparts were added to avoid breaking changes but should not be the case for these new cheatcods

File: crates/cheatcodes/src/test/revert_handlers.rs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                    status,
+                    retdata,
+                    known_contracts,
+                    assume_no_revert.reverted_by.as_ref(),
+                )
+                .ok()
+            })
+            .ok_or_else(|| retdata.clone().into())
+    })
+}
```
Comment (grandizzy) [inline crates/cheatcodes/src/test/revert_handlers.rs:144]:
```suggestion
}

```

File: crates/cheatcodes/src/test/assume.rs [hunk 8]
```diff
@@ -1,12 +1,37 @@
...
+    "Cannot combine a generic assumeNoRevert with specific assumeNoRevert reasons";
+
 #[derive(Clone, Debug)]
 pub struct AssumeNoRevert {
     /// The call depth at which the cheatcode was added.
     pub depth: u64,
+    /// Acceptable revert parameters for the next call, to be thrown out if they are encountered;
+    /// reverts with parameters not specified here will count as normal reverts and not rejects
+    /// towards the counter.
+    pub reasons: Option<Vec<AcceptableRevertParameters>>,
```
Comment (grandizzy) [inline crates/cheatcodes/src/test/assume.rs:21]: I think we can remove the Option and test for empty vec instead

**Review by grandizzy — CHANGES_REQUESTED** [review]
thank you, left couple of comments.

File: crates/cheatcodes/spec/src/vm.rs (hunk 6 above)
Comment (emo-eth) [reply crates/cheatcodes/spec/src/vm.rs:844]: @grandizzy just to make sure I understand – are you saying that `vm.assumeNoRevert(bytes4)` should behave like `assumeNoPartialRevert(bytes4)`?

File: crates/forge/tests/cli/test_cmd.rs [hunk 9]
```diff
@@ -2392,6 +2393,210 @@ contract Dummy {
...
+        /// @dev Test that `assumeNoRevert` assumptions are cleared after the first non-cheatcode external call
+        function testMultipleAssumesClearAfterCall_fails(uint256 x) public view {
+            _vm.assumeNoRevert(Reverter.MyRevert.selector);
+            _vm.assumeNoPartialRevert(Reverter.RevertWithData.selector, address(reverter));
+            reverter.twoPossibleReverts(x);
+    
+            reverter.twoPossibleReverts(2);
+        }
+ 
+        // /// @dev Test that `assumeNoRevert` correctly rejects any error selector when no selector is provided
```
[the remark this answers is not in the document]
Comment (emo-eth) [reply crates/forge/tests/cli/test_cmd.rs:2514]: nope, forgot to clean up, ty!

File: crates/cheatcodes/spec/src/vm.rs (hunk 6 above)
Comment (grandizzy) [reply crates/cheatcodes/spec/src/vm.rs:844]: Yeah, that's my take, wdyt?

Comment (emo-eth) [reply crates/cheatcodes/spec/src/vm.rs:844]: i like it! makes it cleaner.

File: crates/cheatcodes/src/test/assume.rs [hunk 10]
```diff
@@ -20,10 +45,87 @@ impl Cheatcode for assumeCall {
...
+    }
+}
+
+fn assume_no_revert(
+    state: &mut Cheatcodes,
+    depth: u64,
+    reason: Option<Vec<u8>>,
+    reverter: Option<Address>,
+) -> Result {
+    ensure!(state.expected_revert.is_none(), ASSUME_EXPECT_REJECT_MAGIC);
```
Comment (grandizzy) [inline crates/cheatcodes/src/test/assume.rs:105]: I think we can make this impl a little bit nicer like
```rust
    ensure!(state.expected_revert.is_none(), ASSUME_EXPECT_REJECT_MAGIC);

    let params = reason.map(|reason| {
        let partial_match = reason.len() == 4;
        AcceptableRevertParameters { reason, partial_match, reverter }
    });

    match state.assume_no_revert {
        Some(ref mut assume) => {
            ensure!(!assume.reasons.is_empty() && params.is_some(), ASSUME_REJECT_MAGIC);
            assume.reasons.push(params.unwrap());
        }
        None => {
            state.assume_no_revert = Some(AssumeNoRevert {
                depth,
                reasons: if let Some(params) = params { vec![params] } else { vec![] },
                reverted_by: None,
            });
        }
    }

    Ok(Default::default())
```

**Review by grandizzy — COMMENTED** [review]
Thank you, looks good! I pushed a change to simplify test and have one more nit re `assume_no_revert` fn. @zerosnacks @yash-atreya @klkvr  pls chime in when time permits. thank you!

**Review by grandizzy — APPROVED** [review]
Lgtm, thank you! Pending additional review before merging

**Comment by mds1** [comment]
In general UX LGTM! One thing is let's make sure the behavior between this and the various `expectRevert` cheats is consistent, e.g. if you pass a 4byte selector does it only match if the revert data is exact, or do partials match also? Given that this cheat is `assumeNoRevert` for partial match I think it's inconsistent as spec'd in the PR description

https://github.com/foundry-rs/forge-std/blob/da591f56d8884c5824c0c1b3103fbcfd81123c4c/src/Vm.sol#L2102-L2124

**Comment by emo-eth** [comment]
> In general UX LGTM! One thing is let's make sure the behavior between this and the various `expectRevert` cheats is consistent, e.g. if you pass a 4byte selector does it only match if the revert data is exact, or do partials match also? Given that this cheat is `assumeNoRevert` for partial match I think it's inconsistent as spec'd in the PR description
> 
> https://github.com/foundry-rs/forge-std/blob/da591f56d8884c5824c0c1b3103fbcfd81123c4c/src/Vm.sol#L2102-L2124

apologies, i can update the description (pending final decision) – @grandizzy suggested having the default behavior to accept partial matches. 

having the granularity of `assumeNoRevert` vs `assumeNoPartialRevert` would be useful in the scenario of colliding selectors – but that's pretty rare unless intentional. i am having a hard time thinking of cases where a test might be pulling in (arbitrary) external or random revert bytes that might have collisions – but there are possibly some, and it might be worth having the disambiguation. 

i do appreciate the explicitness, though i do think `assumeNoRevert` is already kind of a counter-intuitive name, and `assumeNoPartialRevert` adds onto that complexity. eg `assumeNoRevertPartial` is maybe slightly less confusing, but breaks with the `expectPartialRevert` naming convention...

happy to add back in if there's consensus

**Comment by mds1** [comment]
Thanks for those details, so my thoughts are:
- We already have `expectRevert(bytes4)` and `expectPartialRevert(bytes4)`. To this point—"would be useful in the scenario of colliding selectors – but that's pretty rare unless intentional"—I'm unsure of the motivation for this split
- The least surprising assume behavior would therefore be analogous meaning `assumeNoRevert(bytes4)` and `assumeNoPartialRevert(bytes4)`
- It sounds like there are some breaking change considerations here, which I am not fully up to speed on, in which case that is suitable rationale for having inconsistent behavior between the `expect` and `assume` cheats
- Therefore I'll defer to you and @grandizzy to decide
- But, we should track this and unify / simplify for foundry v1

**Comment by grandizzy** [comment]
thanks both for comments! the `expectPartialRevert` was added when `expectRevert` was already in place and used see https://github.com/foundry-rs/foundry/issues/3725#issue-1457771816 and Paul's comment
> This may warrant a new VM cheatcode, to ensure backwards compatibility and avoid triggering false positives for people who assume the current vm.expectRevert cheatcode to look for an exact ABI match.

For `assumeNoRevert` there's no breaking change consideration (as we don't have them yet) that's why I proposed to avoid adding partial one but consistency is a good point too. Would like to hear more opinions, @zerosnacks maybe you could share your thoughts?

**Comment by zerosnacks** [comment]
Thanks @emo-eth for your PR!

>Is the behavior of calling vm.assumeNoRevert multiple times to build up anticipated reverts per-call acceptable, or should it be forced to take an array of structs with revert parameters, since I think aggregating parameters from separate calls is new behavior (correct me if I'm wrong).

I would prefer we use an array parameter to express the intent of checking against multiple cases in line 
with existing behavior (`mockCalls`, `makePersistent`, `revokePersistent`, etc..). Stateful stacking as proposed is, as mentioned, a new kind of behavior I think we should avoid.

I agree that `expectPartialRevert` exists for a backwards compatibility reason, no need to introduce a parallel `assumeNoPartialRevert` in my opinion. This can always be added later on in a non-breaking way if there is a specific request for it.

>Should these functions be pure and Safe? I see that expectRevert is not pure, and is marked as Unsafe, but assumeNoRevert() is marked pure and Safe.

`Unsafe` generally refers to either it persistently touching EVM settings / state or writing to a file. Do you have an opinion in this regard @DaniPopes?

**Comment by emo-eth** [comment]
@zerosnacks @grandizzy apologies for delay here. Here's a rethought interface that takes an array of structs, let me know your thoughts:

```solidity
    /// Represents an "allowed" revert reason from a single subsequent call when using vm.assumeOnlyAllowedReverts.
    /// Reverts that match will result in a FOUNDRY::ASSUME rejection, whereas unmatched reverts will be surfaced
    /// as normal.
    struct AllowedRevert {
        /// The allowed origin of the revert opcode; address(0) allows reverts from any address
        address reverter;
        /// When true, only matches on the beginning of the revert data, otherwise, matches on entire revert data
        bool partialMatch;
        /// The data to use to match encountered reverts
        bytes revertData;
    }

    /// Discard this run's fuzz inputs and generate new ones if next call reverts with the allowed revert parameters
    #[cheatcode(group = Evm, safety = Safe)]
    function assumeOnlyAllowedRevert(AllowedRevert calldata allowedRevert) external pure;

    /// Discard this run's fuzz inputs and generate new ones if next call reverts with the any of the allowed revert parameters
    #[cheatcode(group = Evm, safety = Safe)]
    function assumeOnlyAllowedReverts(AllowedRevert[] calldata allowedReverts) external pure;
```

I really don't like the `assumeNoRevert` name, so went with something different; I think the change in signature calls for a change in name as well.

**Comment by sambacha** [comment]
can this get merged or what is the hold up?

**Comment by emo-eth** [comment]
I've rebased and refactored. Tests are passing except for the original, no-param `assumeNoRevert()`, which seems to be immediately reverting with a `MAGIC_ASSUME` value somehow, since even a basic test with only `vm.assumeNoRevert()` reaches the global reject limit. I haven't set up the vscode rust debugger on this machine yet so wasn't able to take a crack at it; not sure when I'll have time to tinker next.

**Comment by emo-eth** [comment]
@grandizzy @zerosnacks should be ready for another look!

**Comment by emo-eth** [comment]
@zerosnacks @grandizzy would it makes sense to allow combining with `vm.expectRevert` now that there are negative/count assertions rather than discourage its use? "Throw out any runs that produce this error - unless it comes from this specific reverter" could be useful 🤔

File: crates/cheatcodes/src/test/assume.rs [hunk 11]
```diff
@@ -1,12 +1,49 @@
...
+use alloy_primitives::Address;
 use foundry_evm_core::constants::MAGIC_ASSUME;
-use spec::Vm::{assumeCall, assumeNoRevertCall};
+use spec::Vm::{
+    assumeCall, assumeNoRevert_0Call, assumeNoRevert_1Call, assumeNoRevert_2Call, PotentialRevert,
+};
 use std::fmt::Debug;
 
+pub const ASSUME_REJECT_MAGIC: &str =
+    "Cannot combine a generic assumeNoRevert with specific assumeNoRevert reasons";
```
Comment (zerosnacks) [inline crates/cheatcodes/src/test/assume.rs:10]: Nit, for clarity this can be inlined as it is not used elsewhere

File: crates/cheatcodes/src/test/assume.rs [hunk 12]
```diff
@@ -20,10 +57,41 @@ impl Cheatcode for assumeCall {
...
+    fn apply_stateful(&self, ccx: &mut CheatsCtxt) -> Result {
+        let Self { potentialRevert } = self;
+        assume_no_revert(
+            ccx.state,
+            ccx.ecx.journaled_state.depth(),
+            vec![AcceptableRevertParameters::from(potentialRevert)],
+        )
+    }
+}
+impl Cheatcode for assumeNoRevert_2Call {
```
Comment (zerosnacks) [inline crates/cheatcodes/src/test/assume.rs:76]:
```suggestion
}

impl Cheatcode for assumeNoRevert_2Call {
```

File: crates/cheatcodes/src/test/revert_handlers.rs [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+fn stringify(data: &[u8]) -> String {
+    if let Ok(s) = String::abi_decode(data, true) {
+        return s;
+    }
+    if data.is_ascii() {
+        return std::str::from_utf8(data).unwrap().to_owned();
+    }
+    hex::encode_prefixed(data)
+}
+/// Common parameters for expected or assumed reverts. Allows for code reuse.
```
Comment (zerosnacks) [inline crates/cheatcodes/src/test/revert_handlers.rs:34]:
```suggestion
}

/// Common parameters for expected or assumed reverts. Allows for code reuse.
```

**Review by zerosnacks — COMMENTED** [review]
Hey @emo-eth! Thanks for the updated PR, it looks really good - both the implementation and the completeness of the tests

It looks like the CI is failing related to the proposed changes:

https://github.com/foundry-rs/foundry/actions/runs/12635122448/job/35204363739?pr=9179#step:12:361

Would you mind updating the PR's description to reflect the latest state?

**Comment by grandizzy** [comment]
> It looks like the CI is failing related to the proposed changes:
> 
> https://github.com/foundry-rs/foundry/actions/runs/12635122448/job/35204363739?pr=9179#step:12:361
>
 weird, I don't see why this test would fail with proposed changes and cannot reproduce the failure locally either ... :thinking:

**Comment by emo-eth** [comment]
> > It looks like the CI is failing related to the proposed changes:
> > https://github.com/foundry-rs/foundry/actions/runs/12635122448/job/35204363739?pr=9179#step:12:361
> 
> weird, I don't see why this test would fail with proposed changes and cannot reproduce the test failure locally either ... 🤔

Likewise not sure why my changes would affect it - but does seem to consistently be failing for polygon between runs 🤔

**Comment by grandizzy** [comment]
> > > It looks like the CI is failing related to the proposed changes:
> > > https://github.com/foundry-rs/foundry/actions/runs/12635122448/job/35204363739?pr=9179#step:12:361
> > 
> > 
> > weird, I don't see why this test would fail with proposed changes and cannot reproduce the test failure locally either ... 🤔
> 
> Likewise not sure why my changes would affect it - but does seem to consistently be failing for polygon between runs 🤔

I rebased PR with master containing fixes for other flaky tests (was thinking that failure could be an effect of those) but same test failing, checking some more why's that

**Review by grandizzy — APPROVED** [review]
thank you, lgtm! had to rework the failing test and split tested contracts and test contract in different files as they were affecting outcome. waiting for one more review before merging

**Comment by emo-eth** [comment]
@zerosnacks @yash-atreya @klkvr would love to get this merged if there aren't any more blockers

**Review by zerosnacks — APPROVED** [review]
Thanks @emo-eth!

LGTM

File: crates/forge/tests/cli/test.sol
Comment (DaniPopes) [inline crates/forge/tests/cli/test.sol:1]: this looks unused?

Comment (zerosnacks) [reply crates/forge/tests/cli/test.sol:1]: It is used here https://github.com/foundry-rs/foundry/blob/9ac89d4842f23b9f462e3b3aad895ac134128f4f/crates/cheatcodes/src/test/revert_handlers.rs#L95 and here https://github.com/foundry-rs/foundry/blob/9ac89d4842f23b9f462e3b3aad895ac134128f4f/crates/cheatcodes/src/test/revert_handlers.rs#L186 and the helper is internal

perhaps I'm missing something

Comment (grandizzy) [reply crates/forge/tests/cli/test.sol:1]: yeah, tests within `test_cmd.rs` are not importing this file but rather adding the source inline, so this can be removed
